src/tcpinfo_lib.cc

```diff
@@ -24,8 +223,57 @@ TCPDiagnosticsProto TCPInfoParser::ParseNLMsg(
   return proto;
 }
 
-void TCPInfoParser::NLMsgToProto(const struct nlmsghdr* nlh, Protocol protocol,
+void TCPInfoParser::NLMsgToProto(const struct nlmsghdr* nlh,
+                                 Protocol protocol,
                                  TCPDiagnosticsProto* proto) const {
+  struct inet_diag_msg *r = (struct inet_diag_msg*)NLMSG_DATA(nlh);
+
+  ParseInetDiagMsg(r, proto->mutable_inet_diag_msg());
+
+  // This may be overwritten below.
+  proto->set_diag_protocol(protocol);
+
+  auto* rta = (struct rtattr *)(r+1);
+  int len = nlh->nlmsg_len - NLMSG_LENGTH(sizeof(*r));
+  while (RTA_OK(rta, len)) {
+    switch (rta->rta_type) {
+      case INET_DIAG_PROTOCOL:
+        proto->set_diag_protocol(Protocol(*(__u8 *)RTA_DATA(rta)));
```

Comment:
I would be more comfortable if we logged a message if `Protocol(*(__u8 *)RTA_DATA(rta)) != protocol`